### PR TITLE
Fix some cases of TTL expressions

### LINKED ISSUE #14345: TTL GROUP BY / DB::Exception: Column 'toStartOfHour(timestamp)' already exists.
```
CREATE TABLE derived_metrics_local 
(
  timestamp DateTime,
  bytes UInt64
)
ENGINE=SummingMergeTree()
PARTITION BY toYYYYMMDD(timestamp)
ORDER BY (toStartOfHour(timestamp), timestamp)
TTL toStartOfHour(timestamp) + INTERVAL 1 HOUR GROUP BY toStartOfHour(timestamp)
SET bytes=max(bytes);

insert into derived_metrics_local values('2020-01-01 00:00:00', 1);
DB::Exception: Column 'toStartOfHour(timestamp)' already exists.
```

---

And it would be nice to be able to set PK columns in SET as well

` SET bytes = max(bytes), timestamp = toStartOfHour(timestamp) `

because `timestamp` now is set to some max/any? value after ttl

```
CREATE TABLE derived_metrics_local 
(
  timestamp DateTime,
  timestamp_h DateTime materialized toStartOfHour(timestamp),
  bytes UInt64
)
ENGINE=SummingMergeTree()
PARTITION BY toYYYYMMDD(timestamp)
ORDER BY (timestamp_h, timestamp)
TTL toStartOfHour(timestamp) + INTERVAL 1 HOUR GROUP BY timestamp_h
SET bytes=max(bytes);

insert into derived_metrics_local values('2020-01-01 00:00:00', 111);
insert into derived_metrics_local values('2020-01-01 00:59:02', 1);

OPTIMIZE TABLE derived_metrics_local FINAL;

select timestamp, timestamp_h, bytes  from derived_metrics_local

┌───────────timestamp─┬─────────timestamp_h─┬─bytes─┐
│ 2020-01-01 00:59:02 │ 2020-01-01 00:00:00 │   111 │
└─────────────────────┴─────────────────────┴───────┘


00:59:02 is not related to 111
```

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix creation of `TTL` in cases, when its expression is a function and it is the same as `ORDER BY` key.
Now it's allowed to set custom aggregation to primary key columns in `TTL` with `GROUP BY`.
Backward incompatible: For primary key columns, which are not in `GROUP BY` and aren't set explicitly now is applied function `any` instead of `max`, when TTL is expired. Also if you use TTL with `WHERE` or `GROUP BY` you can see exceptions at merges, while making rolling update.

Fixes #14345.

### CODE REVIEW

File: src/DataStreams/TTLBlockInputStream.cpp [hunk 1]
```diff
@@ -153,112 +153,130 @@ void TTLBlockInputStream::readSuffixImpl()
...
+    expression.execute(block_copy);
+    return block_copy.getByName(result_column).column;
+}
 
-    const IColumn * where_result_column = rows_ttl.where_expression ?
-        block.getByName(rows_ttl.where_result_column).column.get() : nullptr;
+void TTLBlockInputStream::executeRowsTTL(Block & block)
+{
+    auto rows_ttl = metadata_snapshot->getRowsTTL();
+    auto ttl_column = extractRequieredColumn(*rows_ttl.expression, block, rows_ttl.result_column);
```
Comment (CurtizJ) [inline src/DataStreams/TTLBlockInputStream.cpp:172]: Actually fix is here. Previously there was a conflict between calculated columns of primary key and column of TTL expression. 

For example, if we have `ORDER BY toStartOfHour(d)` and `TTL toStartOfHour(d) + INTERVAL 1 HOUR`, in block there will be column `toStartOfHour(d)`, and column with the same name will appear in the middle of computations of TTL expression. There is no way to easily check, that one action of expression is already executed.

Comment (alesapin) [reply src/DataStreams/TTLBlockInputStream.cpp:172]: Why not add this message to the comment of `extractRequieredColumn`?)

File: src/DataStreams/TTLBlockInputStream.cpp [hunk 2]
```diff
@@ -365,34 +379,24 @@ void TTLBlockInputStream::removeValuesWithExpiredColumnTTL(Block & block)
...
     {
         auto & new_ttl_info = ttl_info_map[ttl_entry.result_column];
         if (!block.has(ttl_entry.result_column))
-        {
-            columns_to_remove.push_back(ttl_entry.result_column);
             ttl_entry.expression->execute(block);
-        }
 
-        const IColumn * ttl_column = block.getByName(ttl_entry.result_column).column.get();
+        auto ttl_column = extractRequieredColumn(*ttl_entry.expression, block, ttl_entry.result_column);
```
Comment (alesapin) [inline src/DataStreams/TTLBlockInputStream.cpp:392]: Why we use `extractRequieredColumn` if just executed the expression on the previous line?

File: src/DataStreams/TTLBlockInputStream.cpp [hunk 3]
```diff
@@ -153,112 +153,130 @@ void TTLBlockInputStream::readSuffixImpl()
         LOG_INFO(log, "Removed {} rows with expired TTL from part {}", rows_removed, data_part->name);
 }
 
-void TTLBlockInputStream::removeRowsWithExpiredTableTTL(Block & block)
+static ColumnPtr extractRequieredColumn(const ExpressionActions & expression, const Block & block, const String & result_column)
```
Comment (alesapin) [inline src/DataStreams/TTLBlockInputStream.cpp:156]: Slightly confusing name, maybe mention something about expression execution?

File: src/Storages/MergeTree/MergeTreeDataWriter.cpp [hunk 4]
```diff
@@ -84,19 +84,14 @@ void updateTTL(
...
-    bool remove_column = false;
-    if (!block.has(ttl_entry.result_column))
-    {
-        ttl_entry.expression->execute(block);
-        remove_column = true;
-    }
-
-    const auto & current = block.getByName(ttl_entry.result_column);
+    Block block_copy = block;
+    if (!block_copy.has(ttl_entry.result_column))
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeDataWriter.cpp:91]: Very similar logic to `extractRequieredColumn`.

File: src/Storages/TTLDescription.cpp [hunk 5]
```diff
@@ -182,73 +184,41 @@ TTLDescription TTLDescription::getTTLFromAST(
...
                     ErrorCodes::BAD_TTL_EXPRESSION);
 
-
             result.group_by_keys = Names(pk_columns.begin(), pk_columns.begin() + ttl_element->group_by_key.size());
-
             auto aggregations = ttl_element->group_by_aggregations;
 
-            for (size_t i = 0; i < pk_columns.size(); ++i)
+            const auto & primary_key_expressions = primary_key.expression_list_ast->children;
+            for (size_t i = ttl_element->group_by_key.size(); i < primary_key_expressions.size(); ++i)
```
Comment (alesapin) [inline src/Storages/TTLDescription.cpp:212]: Here we just add all unused primary key columns with `any` aggregation function? Need a comment.

**Review by alesapin — COMMENTED** [review]
Fast test failure seems to be related to changes. Also, backward-incompatible change is the change from `max` function to `any`  in `GROUP BY` TTL?

**Comment by CurtizJ** [comment]
> Also, backward-incompatible change is the change from max function to any in GROUP BY TTL?

Yes.

**Comment by CurtizJ** [comment]
Hmm, very interesting exception :)

DB::Exception: Not found column **any(timestamp)** in block. There are only columns: timestamp_h, max(bytes), **any(timestamp)**, toStartOfHour(any(timestamp))

**Comment by CurtizJ** [comment]
Maybe need a test for backward compatibility.

**Review by alesapin — APPROVED** [review]
LGTM, code became much better than before. Let's add some comments at least at the classes level. Maybe explicitly notice in the description what is backward incompatible here.

**Comment by alesapin** [comment]
All failures are known.

**Pull request merged**